tremor-script/src/ast/analyzer.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020-2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+    ArrayPattern, ArrayPredicatePattern, AssignPattern, Expression, ImutExprInt, Pattern,
+    PredicateClause, PredicatePattern, RecordPattern, TestExpr, TuplePattern,
+};
+
+pub(crate) trait Costly {
+    fn cost(&self) -> u64;
+}
+
+impl Costly for TestExpr {
+    fn cost(&self) -> u64 {
+        self.extractor.cost()
+    }
+}
+
+impl<'script, Ex: Expression + 'script> Costly for PredicateClause<'script, Ex> {
+    fn cost(&self) -> u64 {
+        let g = if self.guard.is_some() { 10 } else { 0 };
+        g + self.pattern.cost()
+    }
+}
+
+impl<'script> Costly for Pattern<'script> {
+    fn cost(&self) -> u64 {
+        match self {
+            Pattern::DoNotCare | Pattern::Default => 0,
+            Pattern::Expr(_) => 10,
+            Pattern::Record(r) => r.cost(),
+            Pattern::Array(a) => a.cost(),
+            Pattern::Assign(a) => a.cost(),
+            Pattern::Tuple(t) => t.cost(),
+            Pattern::Extract(e) => e.cost(),
+        }
+    }
+}
+
+impl<'script> Costly for PredicatePattern<'script> {
+    fn cost(&self) -> u64 {
+        match self {
+            PredicatePattern::FieldPresent { .. } | PredicatePattern::FieldAbsent { .. } => 10,
+            PredicatePattern::Bin {
+                rhs: ImutExprInt::Literal(_),
+                ..
+            } => 20,
+            PredicatePattern::Bin { .. } => 100,
+            PredicatePattern::TildeEq { test, .. } => test.cost(),
+            PredicatePattern::RecordPatternEq { pattern, .. } => pattern.cost(),
+            PredicatePattern::ArrayPatternEq { pattern, .. } => pattern.cost(),
+        }
+    }
+}
+
+impl<'script> Costly for RecordPattern<'script> {
+    fn cost(&self) -> u64 {
+        self.fields.iter().map(|f| f.cost() + 100).sum()
+    }
+}
+
+impl<'script> Costly for ArrayPredicatePattern<'script> {
+    fn cost(&self) -> u64 {
+        match self {
+            ArrayPredicatePattern::Ignore => 1,
+            ArrayPredicatePattern::Expr(ImutExprInt::Literal(_))
+            | ArrayPredicatePattern::Expr(_) => 10,
+            ArrayPredicatePattern::Tilde(t) => t.cost(),
+            ArrayPredicatePattern::Record(r) => r.cost(),
+        }
+    }
+}
+
+impl<'script> Costly for ArrayPattern<'script> {
+    fn cost(&self) -> u64 {
+        let s: u64 = self.exprs.iter().map(Costly::cost).sum();
+        s * (self.exprs.len() as u64)
```

Comment:
wouldnt the sum better represent the cost here? After all we iterate over the array pattern and try every match, no? Sounds like a linear thingy.

Reply:
for each pattern in the array we itterate over the entire target array, so 3 patterns on a 10 element array iterate over 30 elements. We don't know how many elements will be there in the final array but a conservative approximation is that if you write `n` different patterns the array you're looking at will have at least `n` elements.

PS: This should be in a comment :D (will copy it in.

Reply:
Got it. I mixed it up with the tuple pattern logic. All good